RoFILib/configuration/Configuration.cpp

```diff
@@ -0,0 +1,496 @@
+
+    for (const auto& rot : a) {
+        if (moved.find(rot.id()) == moved.end())
+            moved[rot.id()] = {false, false, false};
+
+        if (moved[rot.id()][rot.joint()])
+            return false;
+
+        moved[rot.id()][rot.joint()] = true;
+    }
+
+    return true;
+}
+
+/* * * * * * * * * * * * * *
+ * TRANSFORMATION MATRICES *
+ * * * * * * * * * * * * * */
+
+Matrix transformJoint(double alpha, double beta, double gamma) {
+    return rotate(alpha, X) * rotate(gamma, Z) * translate(Z) * rotate(M_PI, Y) * rotate(-beta, X);
+}
+
+Matrix transformConnection(ConnectorId d1, int ori, ConnectorId d2) {
+    static const std::array<Matrix, 3> dockFaceUp = {
+            rotate(M_PI, Z) * rotate(-M_PI/2, Y), // XPlus
+            rotate(M_PI, Z) * rotate(M_PI/2, Y),  // XMinus
+            identity  // ZMinus
+    };
+
+    static const std::array<Matrix, 3> faceToDock = {
+            rotate(M_PI, Z) * rotate(M_PI/2, Y), // XPlus
+            rotate(M_PI, Z) * rotate(-M_PI/2, Y),   // XMinus
+            identity // ZMinus
+    };
+
+    return faceToDock[d1] * rotate(ori * M_PI/2, Z) * translate(-Z) * dockFaceUp[d2] * rotate(M_PI, X);
+}
+
+/* * * * * * * * *
+ * CONFIGURATION *
+ * * * * * * * * */
+
+Module& Configuration::getModule(ID id) {
+    auto it = modules.find(id);
+    if(it == modules.end())
+        throw std::out_of_range("No module with this id: " + std::to_string(id));
+
+    return it->second;
+}
+
+const Module& Configuration::getModule(ID id) const {
+    auto it = modules.find(id);
+    if(it == modules.end())
+        throw std::out_of_range("No module with this id: " + std::to_string(id));
+
+    return it->second;
+}
+
+std::vector<ID> Configuration::getIDs() const {
+    std::vector<ID> ids;
+    ids.reserve(modules.size());
+    for (auto& [id, m] : modules)
+        ids.push_back(id);
+
+    return ids;
+}
+
+std::vector<Edge> Configuration::getEdges(ID id) const {
+    std::vector<Edge> res;
+    for (auto& e : edges.at(id)) {
+        if (e.has_value())
+            res.push_back(e.value());
+    }
+    return res;
+}
+
+std::vector<Edge> Configuration::getEdges(ID id, const std::unordered_set<ID>& exclude) const {
+    std::vector<Edge> res;
+    for (auto& e : edges.at(id)) {
+        if (e.has_value() && exclude.find(e->id2()) == exclude.end())
+            res.push_back(e.value());
+    }
+    return res;
+}
+
+void Configuration::addModule(double alpha, double beta, double gamma, ID id) {
+    modules.emplace(std::piecewise_construct,
+                    std::forward_as_tuple(id),  // args for key
+                    std::forward_as_tuple(alpha, beta, gamma, id));
+    edges.emplace(std::piecewise_construct,
+                    std::forward_as_tuple(id),  // args for key
+                    std::forward_as_tuple());
+    if ((modules.size() == 1) || (id < fixedId)) {
+        //TODO: je toto opravdu potřeba/proč to chceme
+        fixedId = id;
+    }
```

Comment:
I wanted to talk about this -- why we want to change the fixedId to the smallest one? I understand the `modules.size() == 1` part, but really dont see the `id < fixedId` one.

Reply:
When you add the first module (`modules.size() == 1`), `fixedId` is undefined. Therefore you give it an initial value.

Reply:
Yes. But I don't understand the `id < fixedId` part, why is this the wanted behaviour?

Reply:
Viki specified the module with the smallest id is the fixed one.

Reply:
Oh, I've already broke this in my part of code. I use it as indicator "this is the root of the tree". So do we want to keep that? I feel like in future it would be better to let the user decide which module is the fixed in order to optimize recomputing `matrices`.

Reply:
I agree with the future. Leave it as it is right now.